### ISSUE TITLE
Option to have ArgoCD helm charts with meta.helm.sh annotations

### ISSUE THREAD

**Issue by hanseltime** [argoproj/argo-cd#21250]
# Summary

Allow an argocd configuration option that will add meta.helm.sh annotations when provided.

In the `argocd-cm.yaml`, we could do:

```yaml
helm.addMetaHelmAnnotations: true
```

This would then add the annotations:

```yaml
meta.helm.sh/release-name: <release>
meta.helm.sh/release-namespace: <namespace>
```

# Motivation

We use ArgoCD as our main visibility for status and configuration for our developers.  However, we have to set up certain things before Argo is available (i.e. logging, cni, etc.).  To do this, we need to use helm  to deploy charts with the same values that we map into our ArgoCD deployment of the same chart.  This allows us to have something like fluentbit up early (in case argocd has a failure) and then have ArgoCD take over managing that.  In a reverse of that, we use ArgoCD as an upgrade testing UI for us to try tweaking particular settings on a helm chart and then backport them into IAC as needed. 

The ArgoCD testing case is the bigger problem if it creates a new resource (ex: we created a new storageclass by updating a values.yaml), because helm will refuse to install since the created resource does not have the `meta.helm.sh` annotations.  The only way to solve this is to either manually find the new resources and add the correct annotations or to delete the resource and then helm install - both of which make it much harder to keep an automatic CI process going

Since ArgoCD just uses the helm go pkg and it's template function, it would be ideal if helm supported this, however, due to [this discussion](https://github.com/helm/helm/issues/9630), it seems clear that helm will not be providing this functionality.

# Proposal

As the summary above describes, there would be a new configuration value for argocd called: `helm.addMetaHelmAnnotations`.  This would be an opt-in feature so that existing users see no change in functionality (in case they have set up more complex workflows to handle the above described problem).

If the value is provided, when Applications or ApplicationSets create their helm charts with the release name and namespaces, we would update each manifest that was output with the annotation since we have those values at the moment of template rendering anyway.

**Comment by crenshaw-dev** [comment]
According to the linked thread, Helm uses those annotations to communicate to itself its ownership of those resources. It seems unwise for Argo CD to "trick" Helm. Neither Argo CD nor Helm are really designed for shared ownership of entire resources.  I'm sympathetic to the comment "it seems to work, so why not ship it?" I'm just not sure that's enough to justify a first class feature.

Workarounds include writing a CMP or wrapping the chart with Kustomize and using commonAnnotations.

**Comment by hanseltime** [comment]
My understanding of ArgoCD's "render the template and then kubectl apply it" paradigm for helm was that it was more geared to tolerating shared ownership.  There's no mechanism in ArgoCD that respects those ownership annotations so it's already very happy to overwrite something that helm has previously created (unless I'm missing some critical configuration property).

I wouldn't promote this as a default feature by any means, but since Argo is:

1. Not available for the earliest deployments (i.e. a late actor in manifest management during a new deploy)
2. The one trying to share a resource via a "helm implementation that doesn't enforce all helm semantics"

This feels like a natural feature set to stem the problems that Argo's loose ownership model creates.

I have considered writing a separate CMP for this as well, but that feels like a last-ditch effort since that would then require keeping parity with ArgoCD's helm rendering logic since any helm chart can reveal the need for this CMP during a later upgrade and most engineers will want to have the assurance that the helm functionality is the exact same as before, just with annotations.  

As for wrapping in kustomize or even just specifying annotations at the application level, I'd like to avoid the burden on the user of ArgoCD to be certain that a large tooling change presents the same outputs as normal.

If the ArgoCD project definitively does not want to support this, I can work on a CMP.  If this is something that Argo would support as opt-in, I would be willing to contribute it (to ease the burden).

**Comment by crenshaw-dev** [comment]
> I'd like to avoid the burden on the user of ArgoCD to be certain that a large tooling change presents the same outputs as normal.

Can you rephrase? I'm not sure I follow.

> If the ArgoCD project definitively does not want to support this, I can work on a CMP.

So far I personally lean "no," but we should probably discuss in the contributors call.

**Comment by crenshaw-dev** [comment]
Gotcha, I think i follow now. Basically the features behavior is super simple and predictable, so you don't have to worry much about using it to solve this problem.

You bet! Will see what people think on the contributors call and report back.

**Comment by blakepettersson** [comment]
A `--take-ownership` flag was recently added to Helm (helm/helm#13439). It looks like it's only available for `helm install`/`helm upgrade`, but it perhaps makes more sense to also have that flag added for `helm template` as well?

**Comment by crenshaw-dev** [comment]
Sounds like `--take-ownership` might solve the problem even without adding it to the `template` command. OP says "helm will refuse to install" the chart when the ownership annotations are absent. Sounds like specifying `--take-ownership` is the Helm-endorsed way to solve that problem.

**Comment by hanseltime** [comment]
Oh man, this is the exact ideal that the linked discussion said helm wouldn't do!  It looks like this hasn't yet released as a tool (next release is Jan 15), but it would solve this issue with ArgoCD since anyone  deploying early with helm just needs to use `--take-ownership`.  Thanks for finding this!  I'll go back and update the issues that brought me here to hopefully avoid this getting rehashed in the same way in the future (once the release is out).

**Comment by agaudreault** [comment]
Fixed by https://github.com/helm/helm/pull/13439
